### PR TITLE
[IMP] base: simplify the query that checks the access rights

### PR DESCRIPTION
**Pull request by VincentSchippefilt**
before this commit, checking access rights took 2 queries: 1 for group specific rules
and one for generic rules.
after this commit, a single query checks both

This change is important especially for the loading of the menu done in /web,
where we check the security for each model that has a menu, we are saving 25% of the queries

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/addons/base/models/ir_model.py [hunk 1]
```diff
@@ -1765,28 +1765,17 @@ def check(self, model, mode='read', raise_exception=True):
...
-        # We check if a specific rule exists
+        # We check if a specific or generic rule exists
         self._cr.execute("""SELECT MAX(CASE WHEN perm_{mode} THEN 1 ELSE 0 END)
                               FROM ir_model_access a
                               JOIN ir_model m ON (m.id = a.model_id)
-                              JOIN res_groups_users_rel gu ON (gu.gid = a.group_id)
+                              LEFT JOIN res_groups_users_rel gu ON (gu.gid = a.group_id)
                              WHERE m.model = %s
-                               AND gu.uid = %s
+                               AND (a.group_id is null or gu.uid = %s)
```
Comment (fmdl) [inline odoo/addons/base/models/ir_model.py:1774]: little imp : `a.group_id is null` --> `a.group_id IS NULL`

**Review by legalsylvain — COMMENTED** [review]
make sense.

File: odoo/addons/base/models/ir_model.py [hunk 2]
```diff
@@ -1765,28 +1765,17 @@ def check(self, model, mode='read', raise_exception=True):
 
         self.flush(self._fields)
 
-        # We check if a specific rule exists
+        # We check if a specific or generic rule exists
         self._cr.execute("""SELECT MAX(CASE WHEN perm_{mode} THEN 1 ELSE 0 END)
                               FROM ir_model_access a
                               JOIN ir_model m ON (m.id = a.model_id)
-                              JOIN res_groups_users_rel gu ON (gu.gid = a.group_id)
+                              LEFT JOIN res_groups_users_rel gu ON (gu.gid = a.group_id)
```
Comment (william-andre) [inline odoo/addons/base/models/ir_model.py:1772]: My two cents:
```suggestion
                         LEFT JOIN res_groups_users_rel gu ON (gu.gid = a.group_id)
```

Couldn't be using `bool_or` in the select clause better for performances too? I would assume that the `max` needs to check all the rows, but the `bool_or` can stop quicker to do the aggregation

Comment (william-andre) [reply odoo/addons/base/models/ir_model.py:1772]: FYI it was the base back then, but was changed to be compatible with postgresql 7.4
https://github.com/odoo/odoo/commit/8579038ac7029163c9c631a7f6fec14f7b712637

According to the lastest docs (are they up to date?) we are only supporting 10.0 and up
https://www.odoo.com/documentation/15.0/administration/install/install.html#postgresql

**Review by rco-odoo — APPROVED** [review]
Nice!

**Review by william-andre — APPROVED** [review]

**Review by ryv-odoo — APPROVED** [review]

**Comment by VincentSchippefilt** [comment]
robodoo r+

**Review by rousseldenis — APPROVED** [review]

**Pull request closed without merging**